### PR TITLE
PA-471: Add test to take CustomResource backup and restore

### PR DESCRIPTION
**Pull request by kshithijiyer-px**
**What this PR does / why we need it**:
- [x]  Add test `CustomResourceBackupAndRestore` for doing custom backup and restore
- [x]  Add `CreateBackupWithCustomResourceType()` to take Custom Resource Backup. 

**Which issue(s) this PR fixes** (optional)
Closes # PA-471

**Special notes for your reviewer**:
Later on we would need to alter `CreateBackup` function call to do the same and remove `CreateBackupWithCustomResourceType`.

### CODE REVIEW

**Comment by kshithijiyer-px** [comment]
```
STEP: Delete cluster [destination-cluster] in org [default]
INFO[2023-01-26 16:47:16] Verifying : Description : Test completed successfully ? 
INFO[2023-01-26 16:47:16] Actual:PASS, Expected: PASS         
INFO[2023-01-26 16:47:16] --------Test End------                       
INFO[2023-01-26 16:47:16] #Test: CustomResourceBackupAndRestore        
INFO[2023-01-26 16:47:16] #Description: Create custom resource backup and restore  
INFO[2023-01-26 16:47:16] ------------------------                     

• [SLOW TEST:435.421 seconds]
{CustomResourceBackupAndRestore}
/root/torpedo/tests/backup/backup_test.go:4816
  Create custom resource backup and restore
  /root/torpedo/tests/backup/backup_test.go:4847
------------------------------
SSSSSSSSSSS2023-01-26 16:47:16 +0000:[INFO log.go:#184]  Log Dir: /root/torpedo/CleanupBuckets.log
INFO[2023-01-26 16:47:16] --------Test Start------                     
INFO[2023-01-26 16:47:16] #Test: CleanupBuckets                        
INFO[2023-01-26 16:47:16] #Description: Removing buckets               
INFO[2023-01-26 16:47:16] ------------------------                     
INFO[2023-01-26 16:47:16] Running test from file torpedo/tests/common.go, module: CleanupBuckets 
STEP: Delete bucket [aws-kiyer-test-bucket]
INFO[2023-01-26 16:47:16] Verifying : Description : Test completed successfully ? 
INFO[2023-01-26 16:47:16] Actual:PASS, Expected: PASS         
INFO[2023-01-26 16:47:16] --------Test End------                       
INFO[2023-01-26 16:47:16] #Test: CleanupBuckets                        
INFO[2023-01-26 16:47:16] #Description: Removing buckets               
INFO[2023-01-26 16:47:16] ------------------------                     
INFO[2023-01-26 16:47:16] Dashboard URL : http://aetos.pwx.purestorage.com/resultSet/testSetID/81117 

JUnit report was created: /testresults/junit_basic.xml

Ran 1 of 24 Specs in 550.805 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 23 Skipped
PASS
```

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

File: tests/backup/backup_test.go [hunk 1]
```diff
@@ -4812,6 +4812,124 @@ var _ = Describe("{RestoreEncryptedAndNonEncryptedBackups}", func() {
...
+	var cloudCredUID string
+	backupLocationMap := make(map[string]string)
+	var bkpNamespaces []string
+	var clusterUid string
+	var cloudCredName string
+	var clusterStatus api.ClusterInfo_StatusInfo_Status
+	bkpNamespaces = make([]string, 0)
+
+	JustBeforeEach(func() {
+		StartTorpedoTest("CustomResourceBackupAndRestore", "Create custom resource backup and restore", nil, 0)
```
Comment (mkoppal-px) [inline tests/backup/backup_test.go:4833]: Please add the testrail ID instead of `0`

File: tests/backup/backup_test.go [hunk 2]
```diff
@@ -4812,6 +4812,124 @@ var _ = Describe("{RestoreEncryptedAndNonEncryptedBackups}", func() {
...
+		Step("Restoring the backed up application", func() {
+			log.InfoD("Restoring the backed up application")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			for _, namespace := range bkpNamespaces {
+				backupName := fmt.Sprintf("%s-%s", BackupNamePrefix, namespace)
+				restoreName := fmt.Sprintf("%s-%s", restoreNamePrefix, backupName)
+				restoredNameSpace := fmt.Sprintf("%s-%s", namespace, "restored")
+				namespaceMapping[namespace] = restoredNameSpace
+				CreateRestore(restoreName, backupName, namespaceMapping, SourceClusterName, orgID, ctx)
```
Comment (mkoppal-px) [inline tests/backup/backup_test.go:4904]: Safer to handle the error here in case of debugging

File: tests/backup/backup_test.go [hunk 3]
```diff
@@ -4947,6 +5065,77 @@ func UpdateBackup(backupName string, backupuid string, org_id string, cloudCred
 
 }
 
+// CreateBackup creates backup with custom resources
+func CreateBackupWithCustomResourceType(backupName string, clusterName string, bLocation string, bLocationUID string,
+	namespaces []string, labelSelectors map[string]string, orgID string, uid string, preRuleName string,
+	preRuleUid string, postRuleName string, postRuleUid string, resourceType []string ,ctx context.Context) error {
```
Comment (mkoppal-px) [inline tests/backup/backup_test.go:5071]: Minor formatting issue. There should be a space between comma and `ctx`

**Review by mkoppal-px — CHANGES_REQUESTED** [review]
Apart from a few minor comments, it is a solid change 👍

File: tests/backup/backup_test.go [hunk 4]
```diff
@@ -4812,6 +4812,124 @@ var _ = Describe("{RestoreEncryptedAndNonEncryptedBackups}", func() {
...
+	JustAfterEach(func() {
+		defer EndTorpedoTest()
+		ctx, _ := backup.GetAdminCtxFromSecret()
+		log.InfoD("Deleting the deployed apps after the testcase")
+		for i := 0; i < len(contexts); i++ {
+			opts := make(map[string]bool)
+			opts[SkipClusterScopedObjects] = true
+			taskName := fmt.Sprintf("%s-%d", taskNamePrefix, i)
+			err := Inst().S.Destroy(contexts[i], opts)
+			dash.VerifySafely(err, nil, fmt.Sprintf("Verify destroying app %s, Err: %v", taskName, err))
```
Comment (nikolaypopov) [inline tests/backup/backup_test.go:4926]: nit: we already passing the err in `dash.VerifySafely(err...`, we should remove it from the message `fmt.Sprintf("Verify destroying app %s, Err: %v", taskName, err))`

File: tests/backup/backup_test.go [hunk 5]
```diff
@@ -4812,6 +4812,124 @@ var _ = Describe("{RestoreEncryptedAndNonEncryptedBackups}", func() {
...
+				CreateRestore(restoreName, backupName, namespaceMapping, SourceClusterName, orgID, ctx)
+			}
+		})
+		Step("Compare PVCs on both namespaces", func() {
+			log.InfoD("Compare PVCs on both namespaces")
+			for _, namespace := range bkpNamespaces {
+				pvcs, _ := core.Instance().GetPersistentVolumeClaims(namespace, labelSelectors)
+				restoreNamespace := fmt.Sprintf("%s-%s", namespace, "restored")
+				restoredpvcs, _ := core.Instance().GetPersistentVolumeClaims(restoreNamespace, labelSelectors)
+				dash.VerifyFatal(len(pvcs.Items), len(restoredpvcs.Items), "Compare  number of PVCs")
```
Comment (nikolaypopov) [inline tests/backup/backup_test.go:4913]: nit: remove extra space and we might want to add the namespaces we are comparing in the message, good for debugging purposes

File: tests/backup/backup_test.go [hunk 6]
```diff
@@ -4812,6 +4812,124 @@ var _ = Describe("{RestoreEncryptedAndNonEncryptedBackups}", func() {
...
+			}
+		})
+		Step("Taking backup of applications", func() {
+			log.InfoD("Taking backup of applications")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			for _, namespace := range bkpNamespaces {
+				backupName := fmt.Sprintf("%s-%s", BackupNamePrefix, namespace)
+				err = CreateBackupWithCustomResourceType(backupName, SourceClusterName, backupLocation, backupLocationUID, []string{namespace}, nil, orgID, clusterUid, "", "", "", "", []string{"PersistentVolumeClaim"}, ctx)
+				dash.VerifyFatal(err, nil, "Verifying backup creation")
```
[the remark this answers is not in the document]
Comment (kshithijiyer-px) [reply tests/backup/backup_test.go:4891]: Done

File: tests/backup/backup_test.go (hunk 5 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_test.go:4913]: Done

File: tests/backup/backup_test.go (hunk 4 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_test.go:4926]: Done

File: tests/backup/backup_test.go (hunk 2 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_test.go:4904]: Done

File: tests/backup/backup_test.go (hunk 3 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_test.go:5071]: Done

File: tests/backup/backup_test.go (hunk 1 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_test.go]: Done

**Review by nikolaypopov — APPROVED** [review]

**Comment by kshithijiyer-px** [comment]
```
STEP: Delete cluster [destination-cluster] in org [default]
INFO[2023-01-27 05:12:20] Verifying : Description : Test completed successfully ? 
INFO[2023-01-27 05:12:20] Actual:PASS, Expected: PASS         
INFO[2023-01-27 05:12:20] --------Test End------                       
INFO[2023-01-27 05:12:20] #Test: CustomResourceBackupAndRestore        
INFO[2023-01-27 05:12:20] #Description: Create custom resource backup and restore  
INFO[2023-01-27 05:12:20] ------------------------                     

• [SLOW TEST:436.190 seconds]
{CustomResourceBackupAndRestore}
/root/torpedo/tests/backup/backup_test.go:4816
  Create custom resource backup and restore
  /root/torpedo/tests/backup/backup_test.go:4847
------------------------------
SSSSSSSSSSSSSSS2023-01-27 05:12:20 +0000:[INFO log.go:#184]  Log Dir: /root/torpedo/CleanupBuckets.log
INFO[2023-01-27 05:12:20] --------Test Start------                     
INFO[2023-01-27 05:12:20] #Test: CleanupBuckets                        
INFO[2023-01-27 05:12:20] #Description: Removing buckets               
INFO[2023-01-27 05:12:20] ------------------------                     
INFO[2023-01-27 05:12:20] Running test from file torpedo/tests/common.go, module: CleanupBuckets 
STEP: Delete bucket [aws-kiyer-test-bucket]
INFO[2023-01-27 05:12:20] Verifying : Description : Test completed successfully ? 
INFO[2023-01-27 05:12:20] Actual:PASS, Expected: PASS         
INFO[2023-01-27 05:12:20] --------Test End------                       
INFO[2023-01-27 05:12:20] #Test: CleanupBuckets                        
INFO[2023-01-27 05:12:20] #Description: Removing buckets               
INFO[2023-01-27 05:12:20] ------------------------                     
INFO[2023-01-27 05:12:20] Dashboard URL : http://aetos.pwx.purestorage.com/resultSet/testSetID/81579 

JUnit report was created: /testresults/junit_basic.xml

Ran 1 of 24 Specs in 551.347 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 23 Skipped
PASS

You're using deprecated Ginkgo functionality:
=============================================
Ginkgo 2.0 is under active development and will introduce several new features, improvements, and a small handful of breaking changes.
A release candidate for 2.0 is now available and 2.0 should GA in Fall 2021.  Please give the RC a try and send us feedback!
  - To learn more, view the migration guide at https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md
  - For instructions on using the Release Candidate visit https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#using-the-beta
  - To comment, chime in at https://github.com/onsi/ginkgo/issues/711

  You are using a custom reporter.  Support for custom reporters will likely be removed in V2.  Most users were using them to generate junit or teamcity reports and this functionality will be merged into the core reporter.  In addition, Ginkgo 2.0 will support emitting a JSON-formatted report that users can then manipulate to generate custom reports.

  If this change will be impactful to you please leave a comment on https://github.com/onsi/ginkgo/issues/711
  Learn more at: https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#removed-custom-reporters

To silence deprecations that can be silenced set the following environment variable:
  ACK_GINKGO_DEPRECATIONS=1.16.5


Ginkgo ran 1 suite in 9m11
```

**Review by mkoppal-px — APPROVED** [review]
LGTM

**Pull request merged**